Marlin/src/gcode/gcode.cpp

```diff
@@ -999,25 +992,23 @@ void GcodeSuite::process_parsed_command(const bool no_ok/*=false*/) {
  * This is called from the main loop()
  */
 void GcodeSuite::process_next_command() {
-  char * const current_command = queue.command_buffer[queue.index_r];
+  GCodeQueue::CommandLine &command = queue.peek_next_command();
 
-  PORT_REDIRECT(SERIAL_PORTMASK(queue.port[queue.index_r]));
+  PORT_REDIRECT(SERIAL_PORTMASK(command.port));
 
-  #if ENABLED(POWER_LOSS_RECOVERY)
-    recovery.queue_index_r = queue.index_r;
-  #endif
+  TERN_(POWER_LOSS_RECOVERY, recovery.queue_index_r = queue.ring_buffer.index_r);
 
   if (DEBUGGING(ECHO)) {
     SERIAL_ECHO_START();
-    SERIAL_ECHOLN(current_command);
+    SERIAL_ECHOLN(command.buffer);
     #if ENABLED(M100_FREE_MEMORY_DUMPER)
-      SERIAL_ECHOPAIR("slot:", queue.index_r);
-      M100_dump_routine(PSTR("   Command Queue:"), &queue.command_buffer[0][0], &queue.command_buffer[BUFSIZE - 1][MAX_CMD_SIZE - 1]);
+      SERIAL_ECHOLNPAIR("slot:", queue.ring_buffer.index_r, "   Command Queue:");
+      LOOP_L_N(i, BUFSIZE) SERIAL_ECHOLN(queue.ring_buffer.commands[i].buffer);
```

Comment:
The M100 free memory dump should show if any bytes are corrupted in the buffer, so this simple printing of the buffer as strings may step on the M100 author's intent.

Reply:
Right. The initial code was hacky (taking the address of a `[i][j]` array and matching it with the address of a `[m][n]` array).
There is no guarantee that the memory will be contiguous on a 2D array, the compiler is free to align the inner array the way it wants, so the dump could have contained random bytes anyway (those used for padding).
 
So whatever the change I'll do here (I have to loop over the command buffer, since this not a 2D array anymore), I'll break the memory dump output format.  It's a debug output anyway, so I think it's not important if we change it.

I'll restore the hexdump format here.

Reply:
We can trust C / C++ arrays to be contiguous in RAM under our compilers.